### ISSUE TITLE
ST_3DIntersects segfault

### ISSUE THREAD

**Issue by mhugo** [Oslandia/SFCGAL#171]
https://lists.osgeo.org/pipermail/postgis-devel/2018-August/027300.html

The query below segfaults with me on
"POSTGIS="2.5.0rc1dev r16693" [EXTENSION] PGSQL="100"
GEOS="3.7.0beta2-CAPI-1.11.0 e7d54545" SFCGAL="1.3.5" PROJ="Rel. 5.1.0,
June 1st, 2018" GDAL="GDAL 2.4.0dev-3ea4d4ed14-dirty, released 2018/07/09"
LIBXML="2.9.4" LIBJSON="0.12.1" LIBPROTOBUF="1.2.1" RAST (...)"

I don't have another postgis version at hand to test, can others confirm
that?
The polyhedral surface seems valid, removing decimal precision makes the
query run without segfaulting.

```
WITH data as (
SELECT
ST_GeometryFromText(
'POLYGON Z ((
122395.299 489126.697 8.61546664325712,
122389.298 489128.73 8.55588025324629,
122391.489 489135.198 8.5526708028059,
122397.49 489133.165 8.61225719281685,
122395.299 489126.697 8.61546664325712))'
) as geoma,
ST_GeometryFromText(
'POLYHEDRALSURFACE Z (((
122390.998245685 489133.068537491 0,
122391.003145022 489133.066423547 0,
122391.003145022 489133.066423547 10,
122390.998245685 489133.068537491 10,
122390.998245685 489133.068537491 0
)),((
122391.003145022 489133.066423547 0,
122383.269575402 489114.842869866 0,
122383.269575402 489114.842869866 10,
122391.003145022 489133.066423547 10,
122391.003145022 489133.066423547 0
)))'
) as geomb
)
SELECT ST_3DIntersection(geoma, geomb)
FROM data
```

**Comment by lbartoletti** [comment]
FreeBSD 11.2-RELEASE
```
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 POSTGIS="2.4.4 r16526" PGSQL="100" GEOS="3.6.3-CAPI-1.10.3 80c13047" SFCGAL="1.3.5" PROJ="Rel. 5.1.0, June 1st, 2018" GDAL="GDAL 2.2.4, released 2018/03/19" LIBXML="2.9.7" LIBJSON="0.13.1" RASTER
(1 ligne)
```


```
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 0105000080020000000102000080030000006ACBE10C70E1FD40C8880444B4DA1D4124705F66911C21403EDBFB0970E1FD40FC905444B4DA1D41B4200273901C2140A777D0F86FE1FD402AB12E46B4DA1D41C6C944D18A1C21400102000080020000006ACBE10C70E1FD40C8880444B4DA1D4124705F66911C2140B815F4515DE1FD40B2854C3BA9DA1D41D2AE31B6351C2140
(1 ligne)

```

**Comment by lbartoletti** [comment]
Works also for:
`POSTGIS="2.5.0rc1 r16698" [EXTENSION] PGSQL="95" GEOS="3.6.3-CAPI-1.10.3 80c13047" SFCGAL="1.3.5" PROJ="Rel. 5.1.0, June 1st, 2018" GDAL="GDAL 2.2.4, released 2018/03/19" LIBXML="2.9.7" LIBJSON="0.13.1" RASTER`

` POSTGIS="2.5.0rc1 r16698" [EXTENSION] PGSQL="100" GEOS="3.6.3-CAPI-1.10.3 80c13047" SFCGAL="1.3.5" PROJ="Rel. 5.1.0, June 1st, 2018" GDAL="GDAL 2.2.4, released 2018/03/19" LIBXML="2.9.7" LIBJSON="0.13.1" RASTER`

**Comment by mhugo** [comment]
Related to https://trac.osgeo.org/postgis/ticket/4157

@lbartoletti Strange, it could reproduce it on first try with:
```
POSTGIS="3.0.0dev r16844" [EXTENSION] PGSQL="100" GEOS="3.6.2-CAPI-1.10.2 4d2925d6" SFCGAL="1.3.5" PROJ="Rel. 5.1.0, June 1st, 2018" LIBXML="2.9.4" LIBJSON="0.12.1"
```

**Comment by tomvantilburg** [comment]
🎉 great!

**Comment by Algunenano** [comment]
Hi, this is still crashing on my end:

With ASAN:
```jun 12 14:20:15 Mordor postgres[6293]: AddressSanitizer:DEADLYSIGNAL
jun 12 14:20:15 Mordor postgres[6293]: =================================================================
jun 12 14:20:15 Mordor postgres[6293]: ==6315==ERROR: AddressSanitizer: SEGV on unknown address 0x604000010000 (pc 0x7fe3566f9bee bp 0x603000001330 sp 0x7ffe5132efe8 T0)
jun 12 14:20:15 Mordor postgres[6293]: ==6315==The signal is caused by a READ memory access.
jun 12 14:20:15 Mordor postgres[6293]:     #0 0x7fe3566f9bed in __gmpn_copyi_core2 (/usr/lib/libgmp.so.10+0x33bed)
jun 12 14:20:15 Mordor postgres[6293]: AddressSanitizer can not provide additional info.
jun 12 14:20:15 Mordor postgres[6293]: SUMMARY: AddressSanitizer: SEGV (/usr/lib/libgmp.so.10+0x33bed) in __gmpn_copyi_core2
jun 12 14:20:15 Mordor postgres[6293]: ==6315==ABORTING
```

Coredump (no ASAN):
```
#0  __gmpn_copyi_core2 () at tmp-core2_copyi.s:136
#1  0x00007fec673611f9 in __gmpq_set (dest=dest@entry=0x7ffec2b28bb0, src=src@entry=0x55a08037f760) at set.c:45
#2  0x00007fec67cc5c57 in boost::multiprecision::backends::gmp_rational::gmp_rational (o=..., this=0x7ffec2b28bb0) at /usr/include/boost/multiprecision/gmp.hpp:1844
#3  boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>::number (e=..., this=0x7ffec2b28bb0) at /usr/include/boost/multiprecision/number.hpp:52
#4  boost::array<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>, 3ul>::array (this=0x7ffec2b28bb0) at /usr/include/boost/array.hpp:61
#5  0x00007fec67d273b2 in CGAL::VectorC3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >::VectorC3 (this=0x7ffec2b28bb0) at /usr/include/CGAL/Cartesian/Vector_3.h:36
#6  CGAL::Vector_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >::Vector_3 (this=0x7ffec2b28bb0) at /usr/include/CGAL/Vector_3.h:43
#7  CGAL::PointC3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >::PointC3 (this=0x7ffec2b28bb0) at /usr/include/CGAL/Cartesian/Point_3.h:34
#8  CGAL::Point_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >::Point_3 (this=0x7ffec2b28bb0) at /usr/include/CGAL/Point_3.h:40
#9  CGAL::make_array<CGAL::Point_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > > > (b2=..., b1=...) at /usr/include/CGAL/array.h:113
#10 CGAL::SegmentC3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >::SegmentC3 (ep=..., sp=..., this=0x7ffec2b28bb0) at /usr/include/CGAL/Cartesian/Segment_3.h:54
#11 CGAL::CommonKernelFunctors::Construct_segment_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >::operator() (this=0x55a0802bc940, q=..., p=...)
    at /usr/include/CGAL/Kernel/function_objects.h:2254
#12 CGAL::Lazy_rep_3<CGAL::Segment_3<CGAL::Simple_cartesian<CGAL::Interval_nt<false> > >, CGAL::Segment_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >, CGAL::CommonKernelFunctors::Construct_segment_3<CGAL::Simple_cartesian<CGAL::Interval_nt<false> > >, CGAL::CommonKernelFunctors::Construct_segment_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >, CGAL::Cartesian_converter<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> >, CGAL::Simple_cartesian<CGAL::Interval_nt<false> >, CGAL::NT_converter<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>, CGAL::Interval_nt<false> > >, CGAL::Return_base_tag, CGAL::Point_3<CGAL::Epeck>, CGAL::Point_3<CGAL::Epeck> >::update_exact (this=0x55a0802bc940)
    at /usr/include/CGAL/Lazy.h:455
#13 0x00007fec67ecbb03 in CGAL::Lazy_rep<CGAL::Segment_3<CGAL::Simple_cartesian<CGAL::Interval_nt<false> > >, CGAL::Segment_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >, CGAL::Cartesian_converter<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> >, CGAL::Simple_cartesian<CGAL::Interval_nt<false> >, CGAL::NT_converter<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>, CGAL::Interval_nt<false> > > >::exact (this=0x55a0802bc940) at /usr/include/CGAL/Lazy.h:791
#14 CGAL::Lazy<CGAL::Segment_3<CGAL::Simple_cartesian<CGAL::Interval_nt<false> > >, CGAL::Segment_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >, boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>, CGAL::Cartesian_converter<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> >, CGAL::Simple_cartesian<CGAL::Interval_nt<false> >, CGAL::NT_converter<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>, CGAL::Interval_nt<false> > > >::exact (this=<synthetic pointer>)
    at /usr/include/CGAL/Lazy.h:792
#15 CGAL::exact<CGAL::Segment_3<CGAL::Simple_cartesian<CGAL::Interval_nt<false> > >, CGAL::Segment_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > >, boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>, CGAL::Cartesian_converter<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> >, CGAL::Simple_cartesian<CGAL::Interval_nt<false> >, CGAL::NT_converter<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1>, CGAL::Interval_nt<false> > > > (l=<synthetic pointer>...)
    at /usr/include/CGAL/Lazy.h:92
#16 CGAL::Lazy_construction_variant<CGAL::Epeck, CGAL::CommonKernelFunctors::Intersect_3<CGAL::Simple_cartesian<CGAL::Interval_nt<false> > >, CGAL::CommonKernelFunctors::Intersect_3<CGAL::Simple_cartesian<boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, (boost::multiprecision::expression_template_option)1> > > >::operator()<CGAL::Segment_3<CGAL::Epeck>, CGAL::Segment_3<CGAL::Epeck> > (this=<optimized out>, l2=<synthetic pointer>..., l1=<synthetic pointer>...) at /usr/include/CGAL/Lazy.h:1593
#17 CGAL::intersection<CGAL::Epeck> (b=<synthetic pointer>..., a=<synthetic pointer>...) at /usr/include/CGAL/Intersections_3/Segment_3_Segment_3.h:31
#18 SFCGAL::algorithm::selfIntersectsImpl<3> (line=...) at /tmp/yaourt-tmp-raul/aur-sfcgal-git/src/sfcgal/src/algorithm/intersects.cpp:473
#19 0x00007fec67eb71a6 in SFCGAL::algorithm::selfIntersects3D (l=...) at /tmp/yaourt-tmp-raul/aur-sfcgal-git/src/sfcgal/src/algorithm/intersects.cpp:507
#20 0x00007fec67ecd50a in SFCGAL::algorithm::isValid (p=..., toleranceAbs=@0x7ffec2b29850: 1.0000000000000001e-09) at /usr/include/c++/8.3.0/bits/stl_vector.h:948
#21 0x00007fec67ed03e7 in SFCGAL::algorithm::isValid (g=..., toleranceAbs=@0x7ffec2b29850: 1.0000000000000001e-09) at /tmp/yaourt-tmp-raul/aur-sfcgal-git/src/sfcgal/src/algorithm/isValid.cpp:525
#22 0x00007fec67ed0603 in SFCGAL::SFCGAL_ASSERT_GEOMETRY_VALIDITY_ (g=..., ctxt="") at /tmp/yaourt-tmp-raul/aur-sfcgal-git/src/sfcgal/src/algorithm/isValid.cpp:63
#23 0x00007fec67ed0a0b in SFCGAL::SFCGAL_ASSERT_GEOMETRY_VALIDITY (g=...) at /usr/include/c++/8.3.0/bits/char_traits.h:287
#24 0x00007fec67ed0c79 in SFCGAL::SFCGAL_ASSERT_GEOMETRY_VALIDITY_3D (g=...) at /tmp/yaourt-tmp-raul/aur-sfcgal-git/src/sfcgal/src/algorithm/isValid.cpp:108
#25 0x00007fec67eb004e in SFCGAL::algorithm::intersection3D (ga=..., gb=...) at /tmp/yaourt-tmp-raul/aur-sfcgal-git/src/sfcgal/src/algorithm/intersection.cpp:184
#26 0x00007fec6803a261 in sfcgal_geometry_intersection_3d (ga=0x55a080368710, gb=0x55a080368880) at /tmp/yaourt-tmp-raul/aur-sfcgal-git/src/sfcgal/src/capi/sfcgal_c.cpp:736
#27 0x00007fec6833dcbd in sfcgal_intersection3D (fcinfo=0x55a08035acd0) at lwgeom_sfcgal.c:344
#28 0x000055a07e39ca82 in ExecInterpExpr (state=0x55a08035a7d0, econtext=0x55a08035a470, isnull=0x7ffec2b29d5f) at execExprInterp.c:678
#29 0x000055a07e3ae98d in ExecEvalExprSwitchContext (state=0x55a08035a7d0, econtext=0x55a08035a470, isNull=0x96efc04fe74b00cf) at ../../../src/include/executor/executor.h:313
#30 ExecProject (projInfo=<optimized out>) at ../../../src/include/executor/executor.h:347
#31 ExecScan (node=<optimized out>, accessMtd=0x55a07e3d7a00 <CteScanNext>, recheckMtd=0x55a08035a7d0) at execScan.c:201
#32 0x000055a07e3a52dc in ExecProcNode (node=<optimized out>) at ../../../src/include/executor/executor.h:247
#33 ExecutePlan (estate=<optimized out>, planstate=0x55a08035a128, operation=<optimized out>, numberTuples=<optimized out>, direction=<optimized out>, dest=0x55a080358af0, use_parallel_mode=<optimized out>, sendTuples=<optimized out>, execute_once=<optimized out>)
    at execMain.c:1723
#34 standard_ExecutorRun (queryDesc=<optimized out>, direction=<optimized out>, count=0, execute_once=<optimized out>) at execMain.c:364
#35 0x000055a07e543766 in PortalRunSelect (portal=0x55a0802fa778, forward=<optimized out>, count=0, dest=<optimized out>) at pquery.c:932
#36 0x000055a07e54324a in PortalRun (portal=0x55a0802fa778, count=9223372036854775807, isTopLevel=<optimized out>, run_once=<optimized out>, dest=0x55a080358af0, altdest=0x55a080358af0, completionTag=0x7ffec2b29ff0 "") at pquery.c:773
#37 0x000055a07e541e8a in exec_simple_query (
    query_string=0x55a08028e978 "WITH data as (\nSELECT\nST_GeometryFromText(\n'POLYGON Z ((\n122395.299 489126.697 8.61546664325712,\n122389.298 489128.73 8.55588025324629,\n122391.489 489135.198 8.5526708028059,\n122397.49 489133.165 8.61"...) at postgres.c:1145
#38 0x000055a07e53f64e in PostgresMain (argc=<optimized out>, argv=<optimized out>, dbname=<optimized out>, username=<optimized out>) at postgres.c:4178
#39 0x000055a07e4a95e6 in BackendRun (port=0x55a0802b4940) at postmaster.c:4358
#40 0x000055a07e4a8b33 in BackendStartup (port=<optimized out>) at postmaster.c:4030
#41 ServerLoop () at postmaster.c:1707
#42 0x000055a07e4a5615 in PostmasterMain (argc=<optimized out>, argv=0x55a0802892b0) at postmaster.c:1380
#43 0x000055a07e402149 in main (argc=3, argv=0x55a0802892b0) at main.c:228
```

Versions:
- Postgis current master: `POSTGIS="3.0.0alpha3dev r17489" [EXTENSION] PGSQL="110" GEOS="3.8.0dev-CAPI-1.11.0 " SFCGAL="1.3.7" PROJ="Rel. 5.2.0, September 15th, 2018" GDAL="GDAL 2.4.1, released 2019/03/15" LIBXML="2.9.9" LIBJSON="0.13.1" LIBPROTOBUF="1.3.1" WAGYU="0.4.3 (Internal)" (core procs from "3.0.0de
v r17376" need upgrade) TOPOLOGY (topology procs from "3.0.0dev r17376" need upgrade) RASTER (raster procs from "3.0.0dev r17376" need upgrade) (sfcgal procs from "3.0.0dev r17376" need upgrade)`
- SFCGAL current master (865): 4369440
- CGAL: 4.14
- Boost: 1.69
- gmp: 6.1.2
